offchain/dispatcher/src/drivers/context.rs

```diff
@@ -62,98 +64,133 @@ impl Context {
         input: &Input,
         broker: &impl BrokerSend,
     ) -> Result<(), BrokerFacadeError> {
-        broker.enqueue_input(self.inputs_sent_count, input).await?;
+        let input_block_number = input.block_added.number.as_u64();
+        self.finish_epoch_if_needed(input_block_number, broker)
+            .await?;
+
+        broker.enqueue_input(self.inputs_sent, input).await?;
+
         self.metrics
             .advance_inputs_sent
             .get_or_create(&self.dapp_metadata)
             .inc();
-        self.inputs_sent_count += 1;
-        self.last_event_is_finish_epoch = false;
+
+        self.inputs_sent += 1;
+        self.last_input_epoch = Some(self.calculate_epoch(input_block_number));
+        self.last_finished_epoch.map(|last_finished_epoch| {
+            assert!(self.last_input_epoch.unwrap() >= last_finished_epoch)
+        });
+
         Ok(())
     }
 }
 
 impl Context {
-    fn calculate_epoch(&self, timestamp: u64) -> u64 {
-        assert!(timestamp >= self.genesis_timestamp);
-        (timestamp - self.genesis_timestamp) / self.epoch_length
-    }
-
-    // This logic works because we call this function with `event_timestamp` being equal to the
-    // timestamp of each individual input, rather than just the latest from the blockchain.
-    fn should_finish_epoch(&self, event_timestamp: u64) -> bool {
-        if self.inputs_sent_count == 0 || self.last_event_is_finish_epoch {
-            false
-        } else {
-            let current_epoch = self.calculate_epoch(self.last_timestamp);
-            let event_epoch = self.calculate_epoch(event_timestamp);
-            event_epoch > current_epoch
+    fn calculate_epoch(&self, block_number: u64) -> u64 {
+        assert!(block_number >= self.genesis_block);
```

Comment:
This assert uncovers an issue on the `dispatcher` configuration, which uses the `application_deployment_block_number` instead of the `input_box_block_number` to check for valid inputs (`block_number`). This can be seen at https://github.com/cartesi/rollups-node/blob/70c5816ac155189f40e414a2df22d54d005b052d/offchain/dispatcher/src/setup.rs#L83.

We should replace that with the right configuration otherwise inputs sent before the application has been deployed will be considered as invalid by the assertion here.